runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -378,14 +379,20 @@ public ByteBuffer getCurrentKey() {
 
   private void setTimer(WindowedValue<InputT> timerElement, TimerInternals.TimerData timerData) {
     try {
+      LOG.debug("Setting timer: {} {}", timerElement, timerData);
       // KvToByteBufferKeySelector returns the key encoded
       ByteBuffer encodedKey = (ByteBuffer) keySelector.getKey(timerElement);
       // We have to synchronize to ensure the state backend is not concurrently accessed by the
       // state requests
       try {
         stateBackendLock.lock();
         getKeyedStateBackend().setCurrentKey(encodedKey);
-        timerInternals.setTimer(timerData);
+        if (timerData.getTimestamp().isAfter(GlobalWindow.TIMESTAMP_MAX_VALUE)) {
```

Comment:
Should this be `MAX_TIMESTAMP` (`Long.MAX_VALUE`)? There can be windows between `GlobalWindow.MAX_TIMEMSTAMP` and `MAX_TIMESTAMP` because the former is a standard day before the latter. Actually, `MAX_TIMESTAMP` would be good as a value because triggers using it as a timestamp can never fire because the Watermark cannot go past `MAX_TIMESTAMP`.

Reply:
The constant is actually defined in BoundedWindow https://github.com/apache/beam/blob/a9723d954a92391e9ca79130a199a8eaf38953f2/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/BoundedWindow.java#L57

I will change this to avoid confusion.